### ISSUE TITLE
Veracrypt crypttab option to include PIM count

### ISSUE THREAD

**Issue by JohannesBe** [systemd/systemd#12651]
**Is your feature request related to a problem? Please describe.**
In 2016, the `tcrypt-veracrypt` option [was added](https://github.com/systemd/systemd/pull/4501/files) to crypttab. I have glanced through the crypttab-code, but alas I'm unable to find a setting to provide the [VeraCrypt PIM](https://www.veracrypt.fr/en/Personal%20Iterations%20Multiplier%20%28PIM%29.html).

**Describe the solution you'd like**
Add a `veracrypt-pim` option for the crypttab file. I've also taken a short look at the tcrypt code, and it seems that [here](https://gitlab.com/cryptsetup/cryptsetup/blob/master/lib/libcryptsetup.h#L516) the option can simply be passed using the `veracrypt_pim` param. I think the referenced struct corresponds with the systemd alternative at [line 437](https://github.com/systemd/systemd/blob/36dbf9717030d0a9d69df2e6180284834fa8a851/src/cryptsetup/cryptsetup.c#L437), but am not sure.

In short, I think something similar to the original [tcrypt-veracrypt](https://github.com/systemd/systemd/pull/4501/files) pull request could be done for the PIM, but since I am not a C-expert, not familiar with the code-base and don't know how to compile/test these things I am hesitant to write a PR myself.

P.S. now that we are at it, it might be useful to check if all other options mentioned in the tcrypt struct are present/could be added as well?

**Comment by lifehome** [comment]
I second this feature request, as there should be a prompt/argument for PIM other than just a password.

**Comment by JohannesBe** [comment]
The PIM prompt is another issue I believe, but indeed it should also be possible to have this asked as an input parameter. I'm not entirely sure that this would be easy however, since this is usually configured through the `HOOK` parameter in `mkinitcpio.conf` (by adding something like `encrypt lvm2`). This means that the `encrypt` hook is a shared hook between all disk encryptions and would have to be able to know it is decrypting a veracrypt volume.

This might be possible by providing another suffix in the `CMDLINE_LINUX` parameters? E.g. right now we do something like (grub specific):

```
GRUB_CMDLINE_LINUX_DEFAULT="cryptdevice=UUID=xxx-xxx-xxx-xxx-xxx:cryptlvm"
```

Which maybe could be replaced by something like:


```
GRUB_CMDLINE_LINUX_DEFAULT="cryptdevice=UUID=xxx-xxx-xxx-xxx-xxx:veracryptlvm"
```

Alas, I believe this would require veracrypt (or at least a submodule) to be added to the linux kernel (can anyone confirm this? how does this work?), which I don't see happening any time soon.

**Comment by systemofapwne** [comment]
I wonder why this is not yet implemented. Shouldn't be so much work, or is it?
For my systems, I completely had to skip /etc/crypttab and write my own "crypt.sh" shellscript that is invoked by a crontab. That script now can provide the necessary pim, unlock the drives (with a keyfile) and then mount them. This is quite of ugly and detached from what crypttab and fstab should do. But right now, I see no other way.

**Comment by poettering** [comment]
I left some early review comments on your branch. Please submit as PR once you feel it's ready.

**Comment by poettering** [comment]
Closing, since #27548 has been merged.

**Comment by Sithuk** [comment]
The veracrypt site notes that the PIM is security sensitive.

"The PIM is treated like a secret value that must be entered by the user each time alongside the password."
https://www.veracrypt.fr/en/Personal%20Iterations%20Multiplier%20(PIM).html

Are there security implications to including the PIM in /etc/crypttab using the  veracrypt-pim=<option> ?

**Comment by systemofapwne** [comment]
> The veracrypt site notes that the PIM is security sensitive.
> 
> "The PIM is treated like a secret value that must be entered by the user each time alongside the password." https://www.veracrypt.fr/en/Personal%20Iterations%20Multiplier%20(PIM).html
> 
> Are there security implications to including the PIM in /etc/crypttab using the veracrypt-pim= ?

PIM acts as a modifier for the rounds necessary to derive the actual encryption key from a users password. More rounds = more time per key derivation, which mitigates brute force attacks - Especially on weak passwords.

If /etc/crypttab is itself not on an encrypted device the PIM can be easily leaked/extracted.
If it is however on an encrypted partition that needs to be unlocked, permissions should be made such that only root can read it. This will not leak PIM on a cold device but users with root permissions can still read the pim.

I admit, that the PR is not perfect and should be extended to support user-input for PIM during boot. But this was beyond my required feature set and can still be extended to cryptsetup by anyone.
